client/gutenberg/editor/calypsoify-iframe.tsx

```diff
@@ -699,7 +699,9 @@ class CalypsoifyIframe extends Component<
 						/* eslint-enable jsx-a11y/iframe-has-title */
 					) }
 				</div>
-				<EditorMediaModal
+				<AsyncLoad
+					require="post-editor/editor-media-modal"
+					placeholder={ null }
```

Comment:
This will load the Media Modal chunk on every Gutenframe load, even if it's never opened. The frame UI will be rendered faster (the Media Modal download happens in background, not blocking the frame UI), but the Media Modal download is still mostly wasted.

We could do this instead:
```js
{ isMediaModalVisible && <AsyncLoad require="editor-media-modal" /> }
```

Reply:
This is an interesting point of view. I've actually thought about it a little bit before opening the PR.

> the Media Modal download happens in background, not blocking the frame UI

This PR aims to do exactly this - improve the user-perceived performance without introducing any user-perceived regressions. Here we try to avoid needlessly blocking the frame UI but still prepare the media modal so it would be immediately ready for use when it's requested.

FWIW `isMediaModalVisible` will be set to `true` as late as when the user actually triggers the media modal already, and if we start downloading the media modal at that time, the user will have to wait for the media modal to load. Even if it happens generally quickly, it would be a subtle regression that users with slower connections could perceive.

To summarize, I'm more concerned about the user experience than the potentially lost bytes over the network. One could argue that we ship lost bytes to the user all over Calypso, often loaded with the same `AsyncLoad` strategy.

That being said, I'm more inclined to keep it as-is, but I'm happy to discuss further.

Reply:
> FWIW `isMediaModalVisible` will be set to `true` as late as when the user actually triggers the media modal already, and if we start downloading the media modal at that time, the user will have to wait for the media modal to load.

This, however, is how async loading works in literally every other case. If you click on "Stats" in sidebar, then, immediately after the click, nothing happens at all in the UI. The `stats` section code is loading. Only after few hundreds of milliseconds, the navigation happens and the UI is updated. If you click "View" in a post popup menu, it takes a while before `async-load-web-preview` loads and shows the modal with preview.

Even the Media Modal itself (its non-editor version) is async-loaded this way, e.g., when [editing the site icon](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/site-settings/site-icon-setting/index.jsx#L247).

So, if we don't perceive this behavior as UX regression anywhere else, then why here?

I won't deny that some UX regression is really there, but it's usually mitigated by:
- preloading the chunk on mouse hover: the site settings button mentioned above does that
- showing a good placeholder. For the media modal, we could create a very simple placeholder that has a similar size and look as the full modal, only with gray rectangles inside.

I believe that implementing one of the above (or both) is superior to always loading the modal chunk.

Reply:
I partially agree with you @jsnajdr. I definitely don't want to deny the obvious performance improvement by not downloading extra bytes. But at the same time, I disagree async loading is always conditional - there are tons of cases where it isn't. A simple search for `<AsyncLoad />` demonstrates that there are more non-conditional cases, actually.

What I'm claiming here though, is that what you're suggesting is related to a change in user experience. And while it's undoubtedly good for performance, as a change it has a different focus from the performance focus that is the primary target of this PR. 

Also, would you say those media modal bytes were less lost if it isn't used at all? I don't think so.

That being said, I think we can consider what you're suggesting a potential improvement, but one that should be separately addressed and separately tested, because of its effect on the user experience. 

Because I'm planning to follow-up on other suggestions you had in the PR, I'm happy to give this one a try, too - in another PR.

How does that sound?

Reply:
It might be useful to have two PRs where it's easy to compare how it behaves in both scenarios? It's sometimes hard to make UX judgments hypothetically instead of just by clicking and trying.

Reply:
Exactly - I'd like to try that separately from this PR. It won't be 2 PRs - it will be testing with production against a PR that conditionalizes the async loading. One step at a time, folks 😉

Reply:
> A simple search for <AsyncLoad /> demonstrates that there are more non-conditional cases, actually.

We have some unconditional `AsyncLoad`s in sidebar, like notices about domains or the site's broken status. It is code that we always want to execute, unconditionally. But we want to defer the execution, because we have something more important to render first.

If we `AsyncLoad` something unconditionally, and show it conditionally, then I think it's a bug that should be addressed 🙂 In fact, we have many places in Calypso that use a performance anti-pattern like this:
```js
<UIComponent isVisible={ isVisible } />
```
instead of
```jsx
{ isVisible && <UIComponent /> }
```

In the first case, the `UIComponent` is instantiated, its Redux `connect` mappings called, only to render nothing in the end. With async loading, it's even worse.

> That being said, I think we can consider what you're suggesting a potential improvement, but one that should be separately addressed and separately tested, because of its effect on the user experience.

OK, let's address it in another PR. This one is still a net improvement despite all my objections.

Here, we show the Media Modal only after user clicks on a button. We should use that signal to load the Media Modal just in time, instead of always loading it. And if the delay between the click and the modal rendering is too big, let's solve it with some kind of placeholder. Or smart preloading.